### PR TITLE
Feature/test progress

### PR DESCRIPTION
**Pull request by hellais**
This implements: https://github.com/TheTorProject/ooni-probe/issues/613 https://github.com/TheTorProject/ooni-probe/issues/586

### CODE REVIEW

**Comment by bassosimone** [comment]
```
$ cat .git/refs/heads/feature/test-progress 
d2a1e98e2b178b58f46cde81b116eea5aa7f8e24
$ git diff 8c9f4ec1f49653c8b8197d67e0e4c7850b49430f
$ 
```

File: ooni/nettest.py [hunk 1]
```diff
@@ -541,6 +561,13 @@ def doneReport(self, report_results):
         The state for the NetTest is informed of the fact that another task has
         reached the done state.
         """
+        self._completedInputs += 1
+        log.msg("Status")
+        log.msg("------")
+        log.msg("%d completed %d remaining" % (self._completedInputs,
+                                               self._totalInputs))
+        log.msg("%0.1f%% (ETA: %ds)" % (self.completionPercentage * 100,
+                                        self.completionEta))
```
Comment (bassosimone) [inline ooni/nettest.py]: There is something wrong in the way in which ETA is computed. From my console logs:

```
Status
------
4 completed 0 remaining
0.0% (ETA: -4s)
```

Note: the test here was http-invalid-request-line.

File: ooni/nettest.py [hunk 2]
```diff
@@ -730,6 +761,8 @@ class NetTestCase(object):
 
     localOptions = {}
 
+    _totalInputs = -1
```
Comment (bassosimone) [inline ooni/nettest.py]: I think `None` is a better dummy value here

Comment (bassosimone) [inline ooni/nettest.py]: Or maybe zero? (That is, I don't get why -1 is used here.)

File: ooni/ui/web/server.py [hunk 3]
```diff
@@ -472,7 +472,7 @@ def api_nettest_start(self, request, test_name):
                 400, 'Missing required option: "{}"'.format(option_name)
             )
 
-        except usage.UsageError:
+        except usage.UsageError as ue:
             raise WebUIError(
```
Comment (bassosimone) [inline ooni/ui/web/server.py]: Where is this `ue` used?

**Comment by hellais** [comment]
> the progress is not updated in the #/measurements view

By this you mean that you don't see the progress counter increasing if you just stay on the page and don't refresh? That is a known limitation and is a consequence of the fact that we don't really have a way of broadcasting updates for specific elements in the REST API to the GUI.

The ideal way to do this, I think, would be using something like websockets. An alternative would be to use an approach similar to how we send notifications from the backend to the client (long polling).

I will look again if there is an easy way to support this in the client and the backend, but otherwise I would say we defer this to future versions.

Note: the refresh button, on the right of the logs button, was added specifically for this reason and an easy way to overcome this limitation.

 > I'd add one more empty line before printing the Status text on the standard output

OK

File: ooni/nettest.py (hunk 2 above)
Comment (hellais) [inline ooni/nettest.py]: Yes you are right. It was meant to be some sort of flag, but I realise now that I actually don't check it. I think at this point I should probably use None and check for None in the above.

**Comment by hellais** [comment]
> the progress is not updated in the #/measurements view

I am going to consider integrating this directly into the #723 pull request. This branch here is now ready for merge.

**Comment by bassosimone** [comment]
> I will look again if there is an easy way to support this in the client and the backend, but otherwise I would say we defer this to future versions.

I understand. I would probably be wise to postpone this specific issue. (Maybe it would be also worth to have a ticket, tho, such that people is not going to open a ticket for us?)

**Comment by bassosimone** [comment]
> This branch here is now ready for merge.

I can't guarantee to be able to do further testing by today. I have read aa9d2a4, however, it it makes sense to me. If waiting for me to do another pass is slowing you down, please go on with merge!

**Comment by hellais** [comment]
> I understand. It would probably be wise to postpone this specific issue. (Maybe it would be also worth to have a ticket, tho, such that people is not going to open a ticket for us?)

Filed a ticket for this inside of #728 

> I can't guarantee to be able to do further testing by today. I have read aa9d2a4, however, and it makes sense to me. If waiting for me to do another pass is slowing you down, please merge!

Ok I am going to proceed with merging this so I can work on tagging a release candidate.

**Pull request merged**